# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset.

In [67]:
import pandas as pd
def load_data_from_google_drive(url):
    url_processed='https://drive.google.com/uc?id=' + url.split('/')[-2]
    df = pd.read_csv(url_processed)
    return df

# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use.

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now.

In [68]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

# Ny sektion

In [69]:
df

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
0,1985,20,1228,81,1328,64,N,0
1,1985,25,1106,77,1354,70,H,0
2,1985,25,1112,63,1223,56,H,0
3,1985,25,1165,70,1432,54,H,0
4,1985,25,1192,86,1447,74,H,0
...,...,...,...,...,...,...,...,...
145284,2016,132,1114,70,1419,50,N,0
145285,2016,132,1163,72,1272,58,N,0
145286,2016,132,1246,82,1401,77,N,1
145287,2016,132,1277,66,1345,62,N,0


# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [70]:
df.head()

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
0,1985,20,1228,81,1328,64,N,0
1,1985,25,1106,77,1354,70,H,0
2,1985,25,1112,63,1223,56,H,0
3,1985,25,1165,70,1432,54,H,0
4,1985,25,1192,86,1447,74,H,0


In [71]:
df.tail()

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
145284,2016,132,1114,70,1419,50,N,0
145285,2016,132,1163,72,1272,58,N,0
145286,2016,132,1246,82,1401,77,N,1
145287,2016,132,1277,66,1345,62,N,0
145288,2016,132,1386,87,1433,74,N,0


We can see the dimensions of the dataframe using the the **shape** attribute

In [72]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [73]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset.

In [74]:
df.describe()

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Numot
count,145289.000000,145289.000000,145289.000000,145289.000000,145289.000000,145289.000000,145289.000000
mean,2001.574834,75.223816,1286.720646,76.600321,1282.864064,64.497009,0.044387
std,9.233342,33.287418,104.570275,12.173033,104.829234,11.380625,0.247819
min,1985.000000,0.000000,1101.000000,34.000000,1101.000000,20.000000,0.000000
25%,1994.000000,47.000000,1198.000000,68.000000,1191.000000,57.000000,0.000000
50%,2002.000000,78.000000,1284.000000,76.000000,1280.000000,64.000000,0.000000
75%,2010.000000,103.000000,1379.000000,84.000000,1375.000000,72.000000,0.000000
max,2016.000000,132.000000,1464.000000,186.000000,1464.000000,150.000000,6.000000


Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [75]:
df.max()

,0
Season,2016
Daynum,132
Wteam,1464
Wscore,186
Lteam,1464
Lscore,150
Wloc,N
Numot,6


Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [76]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score.

In [77]:
df['Lscore'].mean()

np.float64(64.49700940883343)

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [78]:
df['Wscore'].argmax()

np.int64(24970)

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [79]:
df['Season'].value_counts()

,count
Season,
2016,5369
2014,5362
2015,5354
2013,5320
2010,5263
2012,5253
2009,5249
2011,5246
2008,5163


**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [80]:
# unique seasons
df['Season'].nunique()

32

**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [81]:
# winning teams and teams with most wins
df['Wteam'].value_counts()

,count
Wteam,
1181,819
1242,804
1246,765
1314,761
1112,746
...,...
1101,18
1446,14
1118,6


# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [82]:
df.iloc[[df['Wscore'].argmax()]]

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
24970,1991,68,1258,186,1109,140,H,0


Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored.

In [83]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

,Lscore
24970,140


When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [84]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [85]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [86]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

np.int64(140)

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [87]:
df.iloc[:3]

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
0,1985,20,1228,81,1328,64,N,0
1,1985,25,1106,77,1354,70,H,0
2,1985,25,1112,63,1223,56,H,0


In [88]:
df.loc[:3]

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
0,1985,20,1228,81,1328,64,N,0
1,1985,25,1106,77,1354,70,H,0
2,1985,25,1112,63,1223,56,H,0
3,1985,25,1165,70,1432,54,H,0


Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive.

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [89]:
df.loc[df['Wscore'].argmax(), 'Lscore']

np.int64(140)

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get.

In [90]:
df.at[df['Wscore'].argmax(), 'Lscore']

np.int64(140)

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers.

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [91]:
df.sort_values('Lscore').head()

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
100027,2008,66,1203,49,1387,20,H,0
49310,1997,66,1157,61,1204,21,H,0
89021,2006,44,1284,41,1343,21,A,0
85042,2005,66,1131,73,1216,22,H,0
103660,2009,26,1326,59,1359,22,H,0


**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively.

In [92]:
# season

df_sorted_season = df.sort_values('Season')
selected_season_data = df_sorted_season.iloc[100:201, df_sorted_season.columns.get_indexer(['Season', 'Wteam', 'Wscore'])]
display(selected_season_data.head())

,Season,Wteam,Wscore
187,1985,1156,112
186,1985,1155,79
185,1985,1154,93
184,1985,1151,71
183,1985,1149,75


In [93]:
# winning
df_sorted_wteam = df.sort_values('Wteam')
selected_wteam_data = df_sorted_wteam.iloc[100:201, df_sorted_wteam.columns.get_indexer(['Season', 'Wteam', 'Wscore'])]
display(selected_wteam_data.head())

,Season,Wteam,Wscore
101650,2008,1102,72
115469,2011,1102,81
97721,2008,1102,67
12448,1988,1102,57
136264,2015,1102,66


In [94]:
# winning score
df_sorted_wscore = df.sort_values('Wscore')
selected_wscore_data = df_sorted_wscore.iloc[100:201, df_sorted_wscore.columns.get_indexer(['Season', 'Wteam', 'Wscore'])]
display(selected_wscore_data.head())

,Season,Wteam,Wscore
126840,2013,1366,43
11847,1988,1193,43
99013,2008,1460,43
9889,1987,1382,43
122700,2012,1349,43


**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [95]:
# Game w highest winning score from data - extract season and winning team.
highest_wscore_game = selected_wscore_data.loc[selected_wscore_data['Wscore'].idxmax()]
season = highest_wscore_game['Season']
winning_team = highest_wscore_game['Wteam']

print(f"seson w highest winning score game: {season}")
print(f"winning team w highest winning score game: {winning_team}")


seson w highest winning score game: 2014
winning team w highest winning score game: 1296


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [96]:
df[df['Wscore'] > 150]

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
5269,1986,75,1258,151,1109,107,H,0
12046,1988,40,1328,152,1147,84,H,0
12355,1988,52,1328,151,1173,99,N,0
16040,1989,40,1328,152,1331,122,H,0
16853,1989,68,1258,162,1109,144,A,0
17867,1989,92,1258,181,1109,150,H,0
19653,1990,30,1328,173,1109,101,H,0
19971,1990,38,1258,152,1109,137,A,0
20022,1990,40,1116,166,1109,101,H,0
22145,1990,97,1258,157,1362,115,H,0


This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100.

In [97]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot
12046,1988,40,1328,152,1147,84,H,0
12355,1988,52,1328,151,1173,99,N,0
25656,1991,84,1106,151,1212,97,H,0
28687,1992,54,1261,159,1319,86,H,0
35023,1993,112,1380,155,1341,91,A,0
52600,1998,33,1395,153,1410,87,H,0


**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [98]:
# Created column 'ScoreDifference'. Absolute difference between winning and losing scores
# Calculated average score diff
# Filtered DataFrame - include only games where 'ScoreDifference' is greater than the av.
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])
average_score_difference = df['ScoreDifference'].mean()
filtered_df = df[df['ScoreDifference'] > average_score_difference]
display(filtered_df.head())

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot,ScoreDifference
0,1985,20,1228,81,1328,64,N,0,17
3,1985,25,1165,70,1432,54,H,0,16
6,1985,25,1228,64,1226,44,N,0,20
8,1985,25,1260,98,1133,80,H,0,18
10,1985,25,1307,103,1288,71,H,0,32


**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [99]:
# game with highest 'ScoreDifference' from the filtered DF
# Extracted season, winning team, and losing team for that game
highest_score_diff_game = filtered_df.loc[filtered_df['ScoreDifference'].idxmax()]
season = highest_score_diff_game['Season']
winning_team = highest_score_diff_game['Wteam']
losing_team = highest_score_diff_game['Lteam']

print(f"Season of game with highest score difference: {season}")
print(f"Winning team of game with highest score difference: {winning_team}")
print(f"Losing team of game with highest score difference: {losing_team}")

Season of game with highest score difference: 1996
Winning team of game with highest score difference: 1409
Losing team of game with highest score difference: 1341


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group.

In [100]:
df.groupby('Wteam')['Wscore'].mean().head()

,Wscore
Wteam,
1101,78.111111
1102,69.893204
1103,75.839768
1104,75.825944
1105,74.960894


This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [101]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: count, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [102]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array.

In [103]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [104]:
# Grouping DF by 'Season' and calculating ave 'Wscore' each season
average_winning_score_by_season = df.groupby('Season')['Wscore'].mean()
display(average_winning_score_by_season)

,Wscore
Season,
1985,74.723040
1986,74.813640
1987,77.993870
1988,79.773704
1989,81.728511
1990,80.846477
1991,82.733932
1992,79.992004
1993,79.508790


**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [105]:
# Grouping DF by 'Wteam' and finding maximum 'Wscore' for each team across all seasons
max_winning_score_by_team = df.groupby('Wteam')['Wscore'].max()
display(max_winning_score_by_team)

,Wscore
Wteam,
1101,95
1102,111
1103,109
1104,114
1105,114
...,...
1460,136
1461,112
1462,125


**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [106]:
# Grouping by 'Season' and 'Wteam'. Finding average winning score for each team per season
# For each season - found team with highest average winning score
average_winning_score_by_season_team = df.groupby(['Season', 'Wteam'])['Wscore'].mean()
highest_avg_wscore_team_per_season = average_winning_score_by_season_team.groupby('Season').idxmax()

print("Team with the highest average winning score for each season:")
display(highest_avg_wscore_team_per_season.head())

Team with the highest average winning score for each season:


,Wscore
Season,
1985,"(1985, 1328)"
1986,"(1986, 1109)"
1987,"(1987, 1380)"
1988,"(1988, 1258)"
1989,"(1989, 1258)"


**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [107]:
# Grouping by 'Season' and 'Wteam'. Counted the number of wins for each team in each season
wins_per_team_per_season = df.groupby(['Season', 'Wteam']).size().reset_index(name='Wins')
display(wins_per_team_per_season.head())

,Season,Wteam,Wins
0,1985,1102,5
1,1985,1103,9
2,1985,1104,21
3,1985,1106,10
4,1985,1108,19


**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [108]:
# For each season, finding the team w most wins using the DF from the previous task
team_with_most_wins_per_season = wins_per_team_per_season.loc[wins_per_team_per_season.groupby('Season')['Wins'].idxmax()]
display(team_with_most_wins_per_season.head())

,Season,Wteam,Wins
217,1985,1385,27
342,1986,1181,32
818,1987,1424,33
861,1988,1112,31
1323,1989,1328,28


**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [109]:
# Grouping the DF by losing team and finding average losing score for each team.
# Compare this with the average winning score from task 3 and identify teams with a higher average losing score.
average_losing_score_by_team = df.groupby('Lteam')['Lscore'].mean()
display(average_losing_score_by_team.head())

# Defining 'average_winning_score_by_team' within cell for standalone execution
average_winning_score_by_team = df.groupby('Wteam')['Wscore'].mean()

combined_scores = pd.DataFrame({
    'AverageWinningScore': average_winning_score_by_team,
    'AverageLosingScore': average_losing_score_by_team
}).fillna(0)
display(combined_scores.head())

teams_higher_avg_losing_score = combined_scores[combined_scores['AverageLosingScore'] > combined_scores['AverageWinningScore']]

print("Teams with a higher average losing score than winning score:")
if not teams_higher_avg_losing_score.empty:
    display(teams_higher_avg_losing_score)
else:
    print("No teams found with a higher average losing score than winning score.")

,Lscore
Lteam,
1101,60.586207
1102,59.201507
1103,64.117347
1104,64.374317
1105,61.675373


,AverageWinningScore,AverageLosingScore
1101,78.111111,60.586207
1102,69.893204,59.201507
1103,75.839768,64.117347
1104,75.825944,64.374317
1105,74.960894,61.675373


Teams with a higher average losing score than winning score:
No teams found with a higher average losing score than winning score.


In [110]:
# Group by losing team and calculate average losing score.
average_losing_score_by_team = df.groupby('Lteam')['Lscore'].mean()
display(average_losing_score_by_team.head())

,Lscore
Lteam,
1101,60.586207
1102,59.201507
1103,64.117347
1104,64.374317
1105,61.675373


In [111]:
# Combine average winning and losing scores into a single DataFrame.
average_winning_score_by_team = df.groupby('Wteam')['Wscore'].mean()


combined_scores = pd.DataFrame({
    'AverageWinningScore': average_winning_score_by_team,
    'AverageLosingScore': average_losing_score_by_team
}).fillna(0)
display(combined_scores.head())

,AverageWinningScore,AverageLosingScore
1101,78.111111,60.586207
1102,69.893204,59.201507
1103,75.839768,64.117347
1104,75.825944,64.374317
1105,74.960894,61.675373


In [112]:
# Identifying teams w a higher average losing score than win score.
teams_higher_avg_losing_score = combined_scores[combined_scores['AverageLosingScore'] > combined_scores['AverageWinningScore']]

print("Teams with a higher average losing score than winning score:")
if not teams_higher_avg_losing_score.empty:
    display(teams_higher_avg_losing_score)
else:
    print("No teams found with a higher average losing score than winning score.")

Teams with a higher average losing score than winning score:
No teams found with a higher average losing score than winning score.


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [113]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [114]:
# Make empty list to store 'HighScoringGame' values
# Iterate through rows of the DF
# If win score is >100, append 'Yes', otherwise append 'No'
# adding new column; 'HighScoringGame'
high_scoring_games = []
for index, row in df.iterrows():
    if row['Wscore'] > 100:
        high_scoring_games.append('Yes')
    else:
        high_scoring_games.append('No')

df['HighScoringGame'] = high_scoring_games
display(df.head())

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot,ScoreDifference,HighScoringGame
0,1985,20,1228,81,1328,64,N,0,17,No
1,1985,25,1106,77,1354,70,H,0,7,No
2,1985,25,1112,63,1223,56,H,0,7,No
3,1985,25,1165,70,1432,54,H,0,16,No
4,1985,25,1192,86,1447,74,H,0,12,No


**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [115]:
# Initialize a dictionary to store the total games played by each team
# Iterate through each row of the DataFrame
# For each game, increment game count for the winning and losing teams
# Updating dictionary to a series. Sort by games played
games_played_by_team = {}

for index, row in df.iterrows():
    wteam = row['Wteam']
    lteam = row['Lteam']

    games_played_by_team[wteam] = games_played_by_team.get(wteam, 0) + 1
    games_played_by_team[lteam] = games_played_by_team.get(lteam, 0) + 1

# Convert the dictionary to a Series for better display
games_played_series = pd.Series(games_played_by_team)
display(games_played_series.sort_values(ascending=False).head())

,0
1181,1013
1314,1010
1246,999
1257,994
1242,993


**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [116]:
# Initialize an empty dictionary to store the highest score difference for each season
# Iterate through all rows of DF
# For each season, comparing the current games score diff with the stored highest
# Update if higher score diff found, storing game details
# Convert the dictionary to a DF
highest_score_diff_per_season = {}

for index, row in df.iterrows():
    season = row['Season']
    score_diff = row['ScoreDifference']

    if season not in highest_score_diff_per_season or score_diff > highest_score_diff_per_season[season]['ScoreDifference']:
        highest_score_diff_per_season[season] = {
            'ScoreDifference': score_diff,
            'Season': season,
            'Wteam': row['Wteam'],
            'Lteam': row['Lteam'],
            'Wscore': row['Wscore'],
            'Lscore': row['Lscore']
        }


highest_score_diff_df = pd.DataFrame.from_dict(highest_score_diff_per_season, orient='index')
display(highest_score_diff_df.head())

,ScoreDifference,Season,Wteam,Lteam,Wscore,Lscore
1985,60,1985,1361,1288,128,68
1986,84,1986,1314,1264,129,45
1987,73,1987,1155,1118,112,39
1988,68,1988,1328,1147,152,84
1989,70,1989,1242,1135,115,45


Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [117]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [118]:
# Count wins and losses for each team using value_counts
# Sum the wins and losses for each team  - fill missing values with 0
wins_count = df['Wteam'].value_counts()
losses_count = df['Lteam'].value_counts()

total_games_vectorized = (wins_count + losses_count).fillna(0).astype(int)
display(total_games_vectorized.sort_values(ascending=False).head())

,count
1181,1013
1314,1010
1246,999
1257,994
1393,993


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [119]:
# Finding the index of the maximum 'ScoreDifference' for each 'Season'
# Using .loc to select the entire row for each of these indices
highest_score_diff_per_season_vectorized = df.loc[df.groupby('Season')['ScoreDifference'].idxmax()]
display(highest_score_diff_per_season_vectorized.head())

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot,ScoreDifference,HighScoringGame
236,1985,33,1361,128,1288,68,H,0,60,Yes
4731,1986,60,1314,129,1264,45,N,0,84,Yes
8240,1987,51,1155,112,1118,39,H,0,73,Yes
12046,1988,40,1328,152,1147,84,H,0,68,Yes
16677,1989,64,1242,115,1135,45,H,0,70,Yes


# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [120]:
df[['Wscore', 'Lscore']].head()

,Wscore,Lscore
0,81,64
1,77,70
2,63,56
3,70,54
4,86,74


Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks.

In [121]:
df.loc[:, ['Wscore', 'Lscore']].head()

,Wscore,Lscore
0,81,64
1,77,70
2,63,56
3,70,54
4,86,74


Note the difference is the return types when you use brackets and when you use double brackets.

In [122]:
type(df['Wscore'])

pandas.core.series.Series

In [123]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations.

In [124]:
df[0:3]

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot,ScoreDifference,HighScoringGame
0,1985,20,1228,81,1328,64,N,0,17,No
1,1985,25,1106,77,1354,70,H,0,7,No
2,1985,25,1112,63,1223,56,H,0,7,No


Here's an equivalent using iloc

In [125]:
df.iloc[0:3,:]

,Season,Daynum,Wteam,Wscore,Lteam,Lscore,Wloc,Numot,ScoreDifference,HighScoringGame
0,1985,20,1228,81,1328,64,N,0,17,No
1,1985,25,1106,77,1354,70,H,0,7,No
2,1985,25,1112,63,1223,56,H,0,7,No


# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [126]:
df.isnull().sum()

,0
Season,0
Daynum,0
Wteam,0
Wscore,0
Lteam,0
Lscore,0
Wloc,0
Numot,0
ScoreDifference,0
HighScoringGame,0


# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis).
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library.
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y